README
JavaPP is a macro processor that runs in the Java VM and resembles the
C preprocessor in syntax and functionality.  The core text processing is
implemented in python and jython is used for running in the Java VM.


 Grammar
================================

Properties
-----------------

    #define prop property value
    #undefine prop
    ${prop}

Properties can be defined when the processor is executed (see Ant task
description) or during processing using #define.  Properties can also
be removed during processing using #undefine.

Property substitution is performed on the text ${prop} and an error is
raised when 'prop' is not defined.

When defining a property, all text following the property identifier
is considered the value.  For example:

  #define var1 "some string of text"

will cause subsequent substitution to contain the double quote characters
and is NOT the same as...

  #define var1 some string of text

Conditions
-----------------

    #if expression [and|or expression]
    [#elif expression [and|or expression]]
    [#else]
    #endif

Conditions can be nested.  Expressions are combined using the
'and', 'or' operators.

Expressions
-----------------

    operand <operator> operand | [!]defined(prop)

Expressions contain an operator and two operands.  The supported
operators are '==', '!=', '<=', '<', '>=', '>'.  The operand can
be either string literal (surrounded with double quotes), number
(integer or decimal), or property substitution in the form ${prop}.

The special expression 'defined(prop)' is supported for testing
if a given property has been defined.  The alternate syntax
'!defined(prop)' can be used to test if a property is not defined.


 Ant task
================================
To use the ant task, put the javapp.jar file where ant can find it
(eg: ANT_HOME/lib) and add the taskdef to your build script:

<taskdef resource="javapp-defs.xml" />

Or put javapp.jar into your project and specify the classpath in the taskdef:

<taskdef resource="javapp-def.xml" classpath="javapp.jar" />

Attributes
-----------------

srcfile    - input file, either srcfile or nested fileset is required
destfile   - output file, only allowed when srcfile is set
destdir    - destination directory, required for nested source filesets
prefix     - change the macro prefix, default is '#'
inheritall - inherit all properties from the project, default is true
verbose    - show verbose output, default is false

Elements
-----------------

<property> - append key/value pairs to the processors environment
<fileset>  - zero or more filesets of source files


 ToDo
================================
On the list of things to do is:
 * friendlier error reporting
 * show line number for undefined properties
 * the #include directive
 * custom macros written in python


CHANGELOG
2008-06-03 Josh Kropf <josh.kropf@sonamobile.com>

    * javapp.py: removed superfluous return statements from action procedures


2008-06-03 Josh Kropf <josh.kropf@sonamobile.com>

    * version 0.4
    * JavaPpTask.java (execute): use friendly message from javapp exception
    * javapp.py: raise friendlier exception on KeyError/property undefined


2008-06-02 Josh Kropf <josh.kropf@sonamobile.com>

    * JavaPp.java (process): throw exception with descriptive on python error
    * JavaPpException.java: new class, extracts message from PyException
    * JavaPpTask.java (execute): delete file on javapp error
    * javapp.py: set filename in scanner for more descriptive errors


2008-06-02 Josh Kropf <josh.kropf@sonamobile.com>

    * version 0.3
    * JavaPpTask.java (verbose): new property, output verbose logging
      (execute): print source file name in verbose mode
      (execute): check modification time of destination file to optimize loop


2008-06-02 Josh Kropf <josh.kropf@sonamobile.com>

    * version 0.2
    * JavaPpTask.java (execute): create parent directory of destination files
    * javapp.py: added undefine directive


test/input.java
//#define poo This is poo

//#if defined(poo)
System.out.println("poo is ${poo}");
//#else
System.out.println("poo is not defined");
//#endif

//#if defined(pee)
System.out.println("pee is ${pee}");
//#else
System.out.println("pee is not defined");
//#endif


test/input/test1.java
//#if defined(pee)
System.out.println("pee is defined");
//#else
System.out.println("pee is undefined");
//#endif


test/input/test2.java
//#if defined(pee)
System.out.println("pee is defined");

//#if ${pee} == "pee"
System.out.println("(if) pee='${pee}'");
//#elif ${pee} == "poo"
System.out.println("(elif) pee='${pee}'");
//#endif

//#else
System.out.println("pee is undefined");
//#endif


src/java/ca/slashdev/javapp/JavaPp.java
/*
 * Copyright 2008 Josh Kropf
 * 
 * This file is part of javapp.
 * 
 * javapp is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * javapp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with javapp; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ca.slashdev.javapp;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Hashtable;
import java.util.Properties;

import org.python.core.Py;
import org.python.core.PyDictionary;
import org.python.core.PyException;
import org.python.core.PyFile;
import org.python.core.PyObject;
import org.python.util.PythonInterpreter;

public class JavaPp {
   private PythonInterpreter py;
   
   public JavaPp(String prefix, Hashtable<Object, Object> env) {
      Properties props = new Properties();
      props.setProperty("python.home", "javapp.jar");
      props.setProperty("python.cachedir.skip", "true");
      
      PythonInterpreter.initialize(System.getProperties(), props, new String[] {""});
      
      py = new PythonInterpreter();
      py.exec("from javapp import process");
      
      Hashtable<PyObject, PyObject> table = new Hashtable<PyObject, PyObject>();
      PyObject pkey, pval;
      String val;
      
      // jython requires dict's to contain only PyObjects in key and value
      for (Object key : env.keySet()) {
         pkey = Py.java2py(key);
         val = env.get(key).toString();
         
         try {
            // first attempt to parse number as integer
            pval = Py.java2py(Integer.parseInt(val));
         } catch (NumberFormatException ie) {
            try {
               // if not integer, try double
               pval = Py.java2py(Double.parseDouble(val));
            } catch (NumberFormatException de) {
               // if neither integer or double, let jython decide
               pval = Py.java2py(val);
            }
         }
         
         table.put(pkey, pval);
      }
      
      py.set("env", new PyDictionary(table));
      py.set("prefix", prefix);
   }
   
   public void process(File input, File output) throws IOException, JavaPpException {
      FileInputStream in = null;
      FileOutputStream out = null;
      
      try {
         in = new FileInputStream(input);
         out = new FileOutputStream(output);
         
         py.set("infile", new PyFile(in, input.getName()));
         py.set("outfile", new PyFile(out));
         
         py.exec("process(infile, outfile, infile.name, env, prefix)");
      } catch (PyException e) {
         throw new JavaPpException(e);
      } finally {
         if ( in != null ) in.close();
         if ( out != null ) out.close();
      }
   }
}


src/java/ca/slashdev/javapp/JavaPpTask.java
/*
 * Copyright 2008 Josh Kropf
 * 
 * This file is part of javapp.
 * 
 * javapp is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * javapp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with javapp; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ca.slashdev.javapp;

import java.io.File;
import java.io.IOException;
import java.util.Hashtable;
import java.util.Vector;

import org.apache.tools.ant.BuildException;
import org.apache.tools.ant.DirectoryScanner;
import org.apache.tools.ant.Project;
import org.apache.tools.ant.Task;
import org.apache.tools.ant.taskdefs.Property;
import org.apache.tools.ant.types.FileSet;
import org.apache.tools.ant.util.FileUtils;

public class JavaPpTask extends Task {
   private String prefix = "#";
   
   private boolean inheritAll = true;
   
   private boolean verbose = false;
   
   private File destDir;
   
   private File destFile;
   
   private File srcFile;
   
   private Vector<FileSet> resources = new Vector<FileSet>();
   
   private Vector<Property> properties = new Vector<Property>();
   
   public String getPrefix() {
      return prefix;
   }
   
   public void setPrefix(String prefix) {
      this.prefix = prefix;
   }
   
   public boolean isInheritAll() {
      return inheritAll;
   }
   
   public void setInheritAll(boolean inheritAll) {
      this.inheritAll = inheritAll;
   }
   
   public boolean isVerbose() {
      return verbose;
   }
   
   public void setVerbose(boolean verbose) {
      this.verbose = verbose;
   }
   
   public void setDestDir(File destDir) {
      this.destDir = destDir;
   }
   
   public File getDestDir() {
      return destDir;
   }
   
   public void setDestFile(File destFile) {
      this.destFile = destFile;
   }
   
   public File getDestFile() {
      return destFile;
   }
   
   public File getSrcFile() {
      return srcFile;
   }
   
   public void setSrcFile(File srcFile) {
      this.srcFile = srcFile;
   }
   
   public void addFileSet(FileSet fileSet) {
      resources.add(fileSet);
   }
   
   public void addProperty(Property prop) {
      properties.add(prop);
   }
   
   private void validate() throws BuildException {
      if ( resources.size() == 0 ) {
         if ( srcFile == null ) {
            throw new BuildException("srcfile or resource collection must be given");
         }
         
         // srcfile is set, verify dest file or dest dir is set
         if ( destFile == null && destDir == null ) {
            throw new BuildException("destfile required when srcfile is given");
         }
         
         if ( destDir != null ) {
            destFile = new File(destDir, srcFile.getName());
         }
      } else {
         if ( srcFile != null ) {
            srcFile = null;
            log("srcfile ignored since resource collection is given", Project.MSG_WARN);
         }
         
         if ( destFile != null ) {
            destFile = null;
            log("destfile ignored since resource collection is given", Project.MSG_WARN);
         }
         
         // destination directory required when resource collection is given
         if ( destDir == null ) {
            throw new BuildException("destdir required when resource collection is given");
         }
      }
   }
   
   @Override
   public void execute() throws BuildException {
      validate();
      
      Hashtable env;
      if ( inheritAll ) {
         env = getProject().getProperties();
      } else {
         env = new Hashtable();
      }
      
      for (Property prop : properties) {
         env.put(prop.getName(), prop.toString());
      }
      
      JavaPp pp = new JavaPp(prefix, env);
      
      if ( destDir != null && !destDir.exists() ) {
         if ( ! destDir.mkdirs() ) {
            throw new BuildException("unable to create destination directory");
         }
      }
      
      if ( srcFile != null ) {
         File parent = destFile.getParentFile();
         if ( !parent.exists() ) {
            if ( !parent.mkdirs() ) {
               throw new BuildException("unable to create parent directory of destination file");
            }
         }
         
         if ( verbose )
            log(String.format("processing %s", srcFile.getName()));
         
         try {
            pp.process(srcFile, destFile);
         } catch (JavaPpException e) {
            destFile.delete();
            throw new BuildException(e.getMessage());
         } catch (IOException e) {
            throw new BuildException(e);
         }
      } else {
         for (FileSet fs : resources) {
            DirectoryScanner ds = fs.getDirectoryScanner(getProject());
            File baseDir = ds.getBasedir(), src, dest = null, parent;
            
            for (String file : ds.getIncludedFiles()) {
               
               try {
                  src = new File(baseDir, file);
                  dest = new File(destDir, file);
                  parent = dest.getParentFile();
                  
                  if ( !parent.exists() ) {
                     if ( !parent.mkdirs() ) {
                        throw new BuildException("unable to create parent directory of destination file");
                     }
                  }
                  
                  if ( verbose )
                     log(String.format("processing %s", file));
                  
                  if ( dest.exists() ) {
                     if ( FileUtils.getFileUtils().isUpToDate(src, dest) ) {
                        if ( verbose )
                           log("destination file is uptodate, processing skipped");
                        
                        continue;
                     }
                  }
                  
                  pp.process(src, dest);
               } catch (JavaPpException e) {
                  if ( dest != null )
                     dest.delete();
                  
                  throw new BuildException(e.getMessage());
               } catch (IOException e) {
                  throw new BuildException(e);
               }
            }
         }
      }
   }
}


src/java/ca/slashdev/javapp/JavaPpException.java
/*
 * Copyright 2008 Josh Kropf
 * 
 * This file is part of javapp.
 * 
 * javapp is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * javapp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with javapp; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ca.slashdev.javapp;

import org.python.core.PyException;

/**
 * 
 * @author josh.kropf
 */
public class JavaPpException extends Exception {
   public JavaPpException(PyException cause) {
      super(cause.value.__str__().toString());
   }
}


src/python/javapp.py
#
# Copyright 2008 Josh Kropf
# 
# This file is part of javapp.
# 
# javapp is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# javapp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with javapp; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
import re
from StringIO import StringIO
from exceptions import Exception
from Plex import *

class CondState:
    def __init__(self, result):
        self.result = result
        self.closed = False
        self.buffer = StringIO()

    def write(self, s):
        if self.result and not self.closed:
            self.buffer.write(s)

    def next(self, result):
        if self.result:
            self.closed = True
        else:
            self.result = result

    def __eq__(self, other):
        return self.result == other

    def __str__(self):
        if self.result:
            return self.buffer.getvalue()
        else:
            return ""

class stack(list):
    def push(self, obj):
        self.insert(0, obj)

    def pop(self):
        return list.pop(self, 0)

    def peek(self):
        return self[0]

    def isempty(self):
        return len(self) == 0

class PpScanner(Scanner):

    def if_cond(self, text):
        self.cond_stack.push(CondState(self.eval_cond(text)))

    def elif_cond(self, text):
        self.cond_stack.peek().next(self.eval_cond(text))

    def else_cond(self, text):
        self.cond_stack.peek().next(True)

    def end_cond(self, text):
        return self.output(str(self.cond_stack.pop()))

    def def_var(self, text):
        tokens = text.split(None, 1)
        if len(tokens) != 1:
            self.env[tokens[0]] = tokens[1]
        else:
            self.env[tokens[0]] = ""

    def undef_var(self, text):
        del(self.env[text])

    def eval_cond(self, cond):
        cond = self.var_re.sub('self.env["\\1"]', cond)
        cond = self.ndef_re.sub('not self.env.has_key("\\1")', cond)
        cond = self.def_re.sub('self.env.has_key("\\1")', cond)
        return eval(cond)

    def expand_var(self, text):
        if self.cond_stack.isempty():
            return str(self.env[text])
        elif self.cond_stack.peek() == True:
            return self.output(self.env[text])

    def output(self, text):
        if self.cond_stack.isempty():
            return text
        else:
            self.cond_stack.peek().write(text)

    var_re = re.compile('\$\{([^\}]+)\}')
    def_re = re.compile('defined\(([^\)]+)\)')
    ndef_re = re.compile('!defined\(([^\)]+)\)')

    def __init__(self, file, filename, env, prefix):
        pre = Str(prefix)
        
        comp_op  = Str("==") | Str("!=") | Str(">") | Str("<") | Str(">=") | Str("<=")
        logic_op = Str("or") | Str("and")

        zspace = Rep(Any(" \t"))
        space  = Rep1(Any(" \t"))

        letter   = Range("AZaz") | Any("_.")
        digit    = Range("09")
        hexdigit = Range("09AFaf")

        ident  = letter + Rep(letter | digit)
        number = Rep1(digit) | (Str("0x") + Rep1(hexdigit)) | Rep1(digit) + Str(".") + Rep1(digit)
        strlit = Str('"') + Rep(AnyBut('"')) + Str('"')

        var     = Str("${") + ident + Str("}")
        operand = strlit | number | var
        defexpr = Opt(Str("!")) + Str("defined(") + ident + Str(")")
        expr    = (operand + zspace + comp_op + zspace + operand) | defexpr
        cond    = expr + Rep(zspace + logic_op + zspace + expr)

        lexicon = Lexicon([
            (Str("${"), Begin('expand')),
            State('expand', [
                (ident,    PpScanner.expand_var),
                (Str("}"), Begin(''))
            ]),
            (pre + Str("if"), Begin('ifcond')),
            State('ifcond', [
                (space,  IGNORE),
                (cond,   PpScanner.if_cond),
                (Eol,    Begin(''))
            ]),
            (pre + Str("elif"), Begin('elifcond')),
            State('elifcond', [
                (space,  IGNORE),
                (cond,   PpScanner.elif_cond),
                (Eol,    Begin(''))
            ]),
            (pre + Str("else"), PpScanner.else_cond),
            (pre + Str("endif"), PpScanner.end_cond),
            (pre + Str("define"), Begin('defvar')),
            State('defvar', [
                (space, IGNORE),
                (ident + Opt(space + Rep(AnyBut("\n"))), PpScanner.def_var),
                (Eol,   Begin(''))
            ]),
            (pre + Str("undefine"), Begin('undefvar')),
            State('undefvar', [
                (space, IGNORE),
                (ident, PpScanner.undef_var),
                (Eol,   Begin(''))
            ]),
            (AnyChar, PpScanner.output)
        ])

        Scanner.__init__(self, lexicon, file, filename)

        self.env = env
        self.cond_stack = stack()

def process(input, output, filename = "", env = {}, prefix = "#"):
    scanner = PpScanner(input, filename, env, prefix)
    scanner.begin('')

    while 1:
        try:
            token, text = scanner.read()
        except KeyError, key:
            raise Exception("'%s', line %d, char %d: property '%s' not defined"
                             % (scanner.position() + (key,)))
        
        if token is None:
            break
        output.write(token)

#if __name__ == '__main__':
#    import sys
#    process(sys.stdin, sys.stdout)
